GPU: Add fallback when 16-bit formats are not supported

riperiperi opened the pull request:
MoltenVK doesn't support these formats on Rosetta/x86, but curiously they are supported on Arm64. This PR adds a conversion to RGBA8 when 16 bit formats are not supported by the host GPU. Generally when one of them is supported, all of them are, so I don't expect this to trigger on existing desktop GPUs.

I'm not too much of a fan of the floating point conversion for 5bit to 8bit, but it works to get 0-31 into 0-255.

Should fix #3951 , once it makes it into a macos build.

Similar to other converted fallbacks, these textures currently can't be flushed back to memory.

riperiperi commented:
There's an issue with textures that don't have their width aligned to 4 bytes that needs to be resolved.

riperiperi commented:
I fixed the issue. Now it takes into account the number of padding pixels (remainder) are present on each line and skips them. It also affects the existing R4G4 to R4G4B4A4 conversion, where right now any remainder will just make it not do the vectorized path.

gdkchan commented on Ryujinx.Graphics.GAL/Format.cs:
```diff
@@ -448,6 +448,27 @@ public static bool IsRtColorCompatible(this Format format)
+        /// <returns>True if the texture format is 16 bit RGBA format, false otherwise</returns>
+        public static bool Is16BitRGBA(this Format format)
+        {
+            switch (format)
```
`A1B5G5R5Unorm` seems to be missing.

riperiperi replied:
It doesn't have a conversion.

gdkchan commented on Ryujinx.Graphics.Texture/PixelConverter.cs:
```diff
@@ -7,30 +8,161 @@ namespace Ryujinx.Graphics.Texture
-            Span<ushort> outputSpan = MemoryMarshal.Cast<byte, ushort>(output);
+                for (int y = 0; y < height; y++)
+                {
+                    for (int i = 0; i < width; i++)
```
I think calling it `x` makes more sense than calling it `i`.

gdkchan commented on Ryujinx.Graphics.Texture/PixelConverter.cs:
```diff
@@ -7,30 +8,161 @@ namespace Ryujinx.Graphics.Texture
+            ReadOnlySpan<ushort> inputSpan = MemoryMarshal.Cast<byte, ushort>(data);
+            Span<uint> outputSpan = MemoryMarshal.Cast<byte, uint>(output);
+
+            float factor5Bit = 255.499f/31f;
```
```suggestion
            float factor5Bit = 255.499f / 31f;
```

gdkchan commented on Ryujinx.Graphics.Texture/PixelConverter.cs:
```diff
@@ -7,30 +8,161 @@ namespace Ryujinx.Graphics.Texture
+            Span<uint> outputSpan = MemoryMarshal.Cast<byte, uint>(output);
+
+            float factor5Bit = 255.499f/31f;
+            float factor6Bit = 255.499f/63f;
```
```suggestion
            float factor6Bit = 255.499f / 63f;
```

gdkchan commented on Ryujinx.Graphics.Texture/PixelConverter.cs:
```diff
@@ -7,30 +8,161 @@ namespace Ryujinx.Graphics.Texture
+                for (int i = 0; i < width; i++)
+                {
+                    ushort packed = inputSpan[offset++];
+                    uint r = (uint)((packed & 31) * factor5Bit);
```
Isn't more efficient to just replicate the bits? Like by doing `((packed << 3) & 0xf8) | ((packed >> 2) & 7)`. I haven't checked if it matches the float version, but that's usually how I see this conversion being done.

riperiperi replied:
That does sound reasonable.

gdkchan commented on Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -373,6 +373,13 @@ public unsafe Capabilities GetCapabilities()
+                GAL.Format.R5G5B5X1Unorm,
+                GAL.Format.B5G6R5Unorm,
+                GAL.Format.B5G5R5A1Unorm,
+                GAL.Format.R4G4B4A4Unorm);
```
`R4G4B4A4Unorm` maps to `A4B4G4R4UnormPack16Ext`, which is not supported on Apple Silicon. I believe this check might prevent all 16-bit formats from being used there. Maybe more granularity for those capabilities will be required. On a side note, the fact that this format is not supported there might be causing issues on N64 Online, so it might be better to do the conversion to RGBA8 there (although not sure if that would fix the issue).

gdkchan commented on Ryujinx.Graphics.Gpu/Image/Texture.cs:
```diff
@@ -859,13 +859,35 @@ public SpanOrArray<byte> ConvertToHostCompatibleFormat(ReadOnlySpan<byte> data,
             {
                 switch (Format)
                 {
-                    case Format.Bc1RgbaSrgb:
```
Why was that removed?

riperiperi replied:
I don't know how this happened. It was cherry-picked from the mac branch, so maybe something went wrong there.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/Texture.cs:
```diff
@@ -859,13 +859,35 @@ public SpanOrArray<byte> ConvertToHostCompatibleFormat(ReadOnlySpan<byte> data,
             else if (!_context.Capabilities.SupportsR4G4Format && Format == Format.R4G4Unorm)
             {
-                result = PixelConverter.ConvertR4G4ToR4G4B4A4(result);
+                result = PixelConverter.ConvertR4G4ToR4G4B4A4(result, width);
```
Move that to the `else` path? Seems wasteful to do this conversion if the host doesn't support the format.

riperiperi replied:
It is converted from R4G4 to R4G4B4A4, then to R8G8B8A8. There's no straight conversion to RGBA8.

gdkchan commented on Ryujinx.Graphics.GAL/Capabilities.cs:
```diff
@@ -18,6 +18,7 @@ namespace Ryujinx.Graphics.GAL
         public readonly bool SupportsBgraFormat;
         public readonly bool SupportsR4G4Format;
         public readonly bool SupportsSnormBufferTextureFormat;
+        public readonly bool Supports16BitRGBAFormat;
```
I think it should be called `Supports16BitRgbaFormat` for consistency with the other names. Although some of those formats are RGB formats rather than RGBA, so maybe there's a better name for this capability.

gdkchan commented on Ryujinx.Graphics.Texture/PixelConverter.cs:
```diff
@@ -7,30 +8,170 @@ namespace Ryujinx.Graphics.Texture
+                    outputPacked |= (outputPacked >> 5) & 0x00070707;
+
+                    uint a = forceAlpha ? 1 : (packed >> 15);
+                    outputPacked &= a * 0xffffffffu;
```
Why does this force RGB to 0 when alpha is 0? I think it's possible that something will have alpha set to 0 but the shader might still access the RGB information.

riperiperi commented:
I've updated the PR with separated R4G4 capability, and added A1B5G5R5 conversion. I've also changed the conversions to remove FP math. Conversions appear to work (when forced) in Mario Kart 8, NSO N64 emulator. I'm not sure what other games use these formats.

The naming is a bit weird. 16-bit on its own would include R8G8, 16-bit packed would include R4G4... so the only common factor between these formats is that they have 5 bit components. That's how I've named the capability.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/Texture.cs:
```diff
@@ -873,6 +873,9 @@ public SpanOrArray<byte> ConvertToHostCompatibleFormat(ReadOnlySpan<byte> data,
                     case Format.Etc2RgbUnorm:
                         result = ETC2Decoder.DecodeRgb(result, width, height, depth, levels, layers);
                         break;
+                    case Format.A1B5G5R5Unorm:
```
Why is that grouped with ETC2?

riperiperi replied:
Because this has been rebased like 15 times.

gdkchan commented on Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -396,6 +396,17 @@ public unsafe Capabilities GetCapabilities()
                 GAL.Format.Etc2RgbSrgb,
                 GAL.Format.Etc2RgbUnorm);
 
+            bool supports5BitComponentFormat = FormatCapabilities.OptimalFormatsSupport(compressedFormatFeatureFlags,
```
Maybe we need different flags here. `compressedFormatFeatureFlags` assumes that the format won't need to be rendered to or blitted to (as you can't do that with compressed formats), but those restrictions doesn't apply to those packed 16-bit formats. Ultimately it wouldn't make much difference right now as we don't have a fallback path for render when the render target format is not supported, but I think it might be worth taking this into account for capability check.

riperiperi replied:
I think it's possible to render to these textures right now, but it will be stored internally as RGBA8 and can't be flushed.

gdkchan commented on Ryujinx.Graphics.GAL/Capabilities.cs:
```diff
@@ -18,7 +18,9 @@ namespace Ryujinx.Graphics.GAL
         public readonly bool Supports3DTextureCompression;
         public readonly bool SupportsBgraFormat;
         public readonly bool SupportsR4G4Format;
+        public readonly bool SupportsRGBA4Format;
```
For consistency, I think the name should be changed to `SupportsR4G4B4A4Format` or the one above should be changed to `SupportsRg4Format`. Same for the conversion method names.

riperiperi replied:
I've renamed everything to specify bits for each component. I also renamed Is16BitRGBA to is16BitPacked.

gdkchan reviewed (approved):
lgtm, thanks.
